### PR TITLE
Completed implementation for TCPLT-12581

### PR DESCRIPTION
**Pull request by Priyan-96**

### CODE REVIEW

File: file-service-tcps-trigger/file_service_tcps_trigger/file_service_tcps_trigger_stack.py [hunk 1]
```diff
@@ -49,6 +49,33 @@ def __init__(self, scope: Construct, construct_id: str, tc_env: str, env_context
 
         file_topic.add_subscription(subscriptions.SqsSubscription(tcps_trigger_queue, filter_policy=filter_policy))
 
+        # Create DLQ for Folder Delete Revision Queue
+        folder_delete_dlq = sqs.Queue(self, "FolderDeleteRevision-DLQ",
+            retention_period=Duration.days(14),
+            queue_name="FolderDeleteRevisionDLQ" + resource_name_suffix + ".fifo",
```
Comment (tspriyan28) [inline file-service-tcps-trigger/file_service_tcps_trigger/file_service_tcps_trigger_stack.py:55]: Consider adding explicit timeout configuration for the DLQ. Even though the default retention period is set to 14 days, it's beneficial to specify a visibility timeout as well for consistency with the main queue configuration.

File: file-service-tcps-trigger/file_service_tcps_trigger/file_service_tcps_trigger_stack.py [hunk 2]
```diff
@@ -49,6 +49,33 @@ def __init__(self, scope: Construct, construct_id: str, tc_env: str, env_context
...
+        folder_delete_revision_dlq = DeadLetterQueue(max_receive_count=3, queue=folder_delete_dlq)
+
+        # Create SQS FIFO Queue for Folder Delete Revision
+        folder_delete_revision_queue = sqs.Queue(self, "FolderDeleteRevision-Queue",
+            queue_name="FolderDeleteRevisionQueue" + resource_name_suffix + ".fifo",
+            fifo=True,
+            content_based_deduplication=True,
+            dead_letter_queue=folder_delete_revision_dlq,
+            visibility_timeout=Duration.minutes(5)
+        )
```
Comment (tspriyan28) [inline file-service-tcps-trigger/file_service_tcps_trigger/file_service_tcps_trigger_stack.py:67]: The filter policy uses "FOLDER_ASYNC_ACTION" but the message documentation mentions "FOLDER_DELETE_ACTION". Please verify that the right filter policy key is being used to match the expected messages.

File: file-service-tcps-trigger/functions/FolderDeleteRevisionFunction/folder_delete_revision_fn.py [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+    @param context: consists of lambda context information like requestId
+    """
+    # Get AWS lambda request id
+    request_id = context.aws_request_id
+    # Update logger with awsRequestId and json formatting
+    logging_utils.update_logger_format(request_id=request_id)
+    log.info(f'Received message(s): {event} from SQS to process')
+    try:
+        # Perform Folder Delete Revision on each record sequentially
+        for record in event['Records']:
```
Comment (tspriyan28) [inline file-service-tcps-trigger/functions/FolderDeleteRevisionFunction/folder_delete_revision_fn.py:39]: Consider adding logging for the API call result. Currently, a success log is added but there's no logging for the response data, which could be helpful for debugging.

File: file-service-tcps-trigger/functions/FolderDeleteRevisionFunction/folder_delete_revision_service.py [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+        self.__identity_auth = IdentityAuth()
+        self.__tcps_base_url = os.environ['TCPS_BASE_URL']
+
+    def invoke_tcps_api(self, folder_event: FolderDeleteQueueMessage):
+        """
+        Performs Folder Delete Revision for the folder event.
+        :param folder_event: folder service event derived from sns message
+        """
+        folder_id = folder_event.input.folder_id if folder_event else "Unknown"
+        log.info(f'Performing Folder Delete Revision for the folder {folder_id}')
```
Comment (tspriyan28) [inline file-service-tcps-trigger/functions/FolderDeleteRevisionFunction/folder_delete_revision_service.py:25]: The API endpoint is hardcoded as `/tc/api/2.0/files/logRevisionAsync`. Consider making this configurable through an environment variable to support easier maintenance and future changes.

File: file-service-tcps-trigger/functions/FolderDeleteRevisionFunction/folder_delete_revision_service.py [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+        api_endpoint = self.__tcps_base_url + "/tc/api/2.0/files/logRevisionAsync"
+        
+        # Perform the TCPS API call
+        try:
+            api_client.ApiClient.make_request("POST", api_endpoint, request_headers, request_body)
+            log.info(f'Successfully invoked TCPS API for folder delete revision: {folder_id}')
+        except ApiError as api_error:
+            # Enhanced error logging for 403 errors
+            if api_error.response_code == 401:
+                self.__identity_auth.clear_access_token_and_expiry()
```
Comment (tspriyan28) [inline file-service-tcps-trigger/functions/FolderDeleteRevisionFunction/folder_delete_revision_service.py:40]: The error handling for 401 responses is handled well by clearing the cached token, but the error is still raised. Consider implementing a retry mechanism specifically for 401 errors after clearing the token.

File: file-service-tcps-trigger/functions/FolderDeleteRevisionFunction/model/folder_event.py [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+    code: Optional[str] = None
+    message: Optional[str] = None
+
+@dataclass
+class FolderDeleteResult(object):
+    download_url: Optional[str] = None
+    object_to_delete: Optional[StorageObject] = None
+    object_deleted: Optional[StorageObject] = None
+    type: Optional[str] = None
+    parent_storage_object_id: Optional[int] = None
```
Comment (tspriyan28) [inline file-service-tcps-trigger/functions/FolderDeleteRevisionFunction/model/folder_event.py:48]: The `FileDeleteQueueMessage` class includes several optional fields. Consider adding validation to ensure required fields for processing are present before making the API call.

File: file-service-tcps-trigger/functions/FolderDeleteRevisionFunction/utils/api_client.py [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+        super().__init__(self.message)
+
+    def __str__(self):
+        return self.message
+
+
+class ApiClient(object):
+    """Performs http(s) calls and throws ApiError in case response code is not < 300."""
+
+    @staticmethod
```
Comment (tspriyan28) [inline file-service-tcps-trigger/functions/FolderDeleteRevisionFunction/utils/api_client.py:27]: The backoff mechanism is well implemented, but consider adjusting the `max_tries` and `max_time` parameters based on expected API response times. Also, consider adding jitter to prevent thundering herd problems during retries.

File: file-service-tcps-trigger/functions/FolderDeleteRevisionFunction/tests/unit/test_folder_delete_revision_service.py [hunk 8]
```diff
@@ -0,0 +1,185 @@
+import datetime
```
Comment (tspriyan28) [inline file-service-tcps-trigger/functions/FolderDeleteRevisionFunction/tests/unit/test_folder_delete_revision_service.py:1]: The test coverage appears comprehensive, but consider adding specific tests for edge cases such as handling empty or malformed event messages to ensure robust error handling.

**Review by tspriyan-96 — COMMENTED** [review]
I've completed a detailed review of the implementation for TCPLT-12581 which handles folder delete revision logs. Overall, the implementation appears to be well-structured and follows the requirements specified in the JIRA ticket. The PR adds the necessary infrastructure and code to handle folder deletion revision logs by creating a new lambda function, SQS queues, and implementing the API calls to the monolith's new `/logRevisionAsync` endpoint.
